### PR TITLE
feat(compat): v0 query fees support

### PR DESCRIPTION
**Pull request by TheAncientGoat**
This adds support for the v2 queryFees RPC, which is needed for the legacy UI's to keep working as expected

### CODE REVIEW

File: resolver/resolver.go [hunk 1]
```diff
@@ -285,7 +286,65 @@ func (resolver *Resolver) QueryStat(ctx context.Context, id interface{}, params
...
+
+	select {
+	case <-ctx.Done():
+		resolver.logger.Error("timeout when waiting for response: %v", ctx.Err())
+		jsonErr := jsonrpc.NewError(jsonrpc.ErrorCodeInternal, "request timed out", nil)
+		return jsonrpc.NewResponse(id, nil, &jsonErr)
+	case response := <-reqWithResponder.Responder:
+		raw, err := json.Marshal(response.Result)
+		if err != nil {
+			resolver.logger.Errorf("[resolver] error marshaling queryState result: %v", err)
```
Comment (jazg) [inline resolver/resolver.go:306]: We should return a response back to the user here instead of falling through.

File: resolver/resolver.go [hunk 2]
```diff
@@ -285,7 +286,65 @@ func (resolver *Resolver) QueryStat(ctx context.Context, id interface{}, params
...
+		}
+
+		var resp map[string]map[string]map[string]interface{}
+		if err := json.Unmarshal(raw, &resp); err != nil {
+			resolver.logger.Warnf("[resolver] cannot unmarshal queryState result from %v", err)
```
Comment (jazg) [inline resolver/resolver.go:311]: This one shouldn't fall through either. Also, I'm not sure the warning formatting is quite right (`from %v` where `%v` is the error).

File: resolver/resolver.go [hunk 3]
```diff
@@ -285,7 +286,65 @@ func (resolver *Resolver) QueryStat(ctx context.Context, id interface{}, params
...
+		resolver.logger.Error("timeout when waiting for response: %v", ctx.Err())
+		jsonErr := jsonrpc.NewError(jsonrpc.ErrorCodeInternal, "request timed out", nil)
+		return jsonrpc.NewResponse(id, nil, &jsonErr)
+	case response := <-reqWithResponder.Responder:
+		raw, err := json.Marshal(response.Result)
+		if err != nil {
+			resolver.logger.Errorf("[resolver] error marshaling queryState result: %v", err)
+		}
+
+		var resp map[string]map[string]map[string]interface{}
```
Comment (jazg) [inline resolver/resolver.go:309]: Can we unmarshal into a stricter type here?

File: resolver/resolver.go [hunk 4]
```diff
@@ -285,7 +286,65 @@ func (resolver *Resolver) QueryStat(ctx context.Context, id interface{}, params
...
+		if err := json.Unmarshal(raw, &resp); err != nil {
+			resolver.logger.Warnf("[resolver] cannot unmarshal queryState result from %v", err)
+		}
+
+		state := jsonrpc.ResponseQueryState{}
+		state.State = make(map[multichain.Chain]pack.Struct)
+
+		for i := range resp["state"] {
```
Comment (jazg) [inline resolver/resolver.go:317]: Same as the unmarshal point but would prefer if we could do something like `resp.State`, `resp.State[i].GasCap`, etc.

File: resolver/resolver.go [hunk 5]
```diff
@@ -285,7 +286,65 @@ func (resolver *Resolver) QueryStat(ctx context.Context, id interface{}, params
...
+			gasCap, err := strconv.Atoi(resp["state"][i]["gasCap"].(string))
+			if err != nil {
+				resolver.logger.Error("[resolver] missing gasCap for %v", i)
+				jsonErr := jsonrpc.NewError(jsonrpc.ErrorCodeInternal, "failed compatability conversion", nil)
```
Comment (jazg) [inline resolver/resolver.go:321]:
```suggestion
				jsonErr := jsonrpc.NewError(jsonrpc.ErrorCodeInternal, "failed compatibility conversion", nil)
```

File: resolver/resolver.go [hunk 6]
```diff
@@ -285,7 +286,65 @@ func (resolver *Resolver) QueryStat(ctx context.Context, id interface{}, params
...
+				return jsonrpc.NewResponse(id, nil, &jsonErr)
+			}
+
+			gasLimit, err := strconv.Atoi(resp["state"][i]["gasLimit"].(string))
+			if err != nil {
+				resolver.logger.Error("[resolver] missing gasLimit for %v", i)
+				jsonErr := jsonrpc.NewError(jsonrpc.ErrorCodeInternal, "failed compatability conversion", nil)
```
Comment (jazg) [inline resolver/resolver.go:328]:
```suggestion
				jsonErr := jsonrpc.NewError(jsonrpc.ErrorCodeInternal, "failed compatibility conversion", nil)
```

File: resolver/resolver.go [hunk 7]
```diff
@@ -285,7 +286,65 @@ func (resolver *Resolver) QueryStat(ctx context.Context, id interface{}, params
...
+			state.State[multichain.Chain(i)] = pack.NewStruct(
+				"gasCap", pack.NewU64(uint64(gasCap)),
+				"gasLimit", pack.NewU64(uint64(gasLimit)),
+			)
+		}
+
+		shards, err := v0.QueryFeesResponseFromState(state)
+
+		if err != nil {
+			resolver.logger.Error("failed to cast to QueryFees")
```
Comment (jazg) [inline resolver/resolver.go:341]: Not sure how useful this is if we don't print the error. Can probably just remove it entirely.

File: compat/v0/compat.go [hunk 8]
```diff
@@ -68,6 +68,87 @@ func ShardsResponseFromState(state jsonrpc.ResponseQueryState) (ResponseQuerySha
 	return resp, nil
 }
 
+// ShardsResponseFromState takes a QueryState rpc response and converts it into a QueryShards rpc response
+// It can be a standalone function as it has no dependencies
+func QueryFeesResponseFromState(state jsonrpc.ResponseQueryState) (ResponseQueryFees, error) {
+	bitcoinCap, ok := state.State[multichain.Bitcoin].Get("gasCap").(pack.U64)
+	if !ok {
+		fmt.Printf("state: %v", state.State)
```
Comment (jazg) [inline compat/v0/compat.go:76]:
```suggestion
```

File: compat/v0/compat.go [hunk 9]
```diff
@@ -68,6 +68,87 @@ func ShardsResponseFromState(state jsonrpc.ResponseQueryState) (ResponseQuerySha
...
+		return ResponseQueryFees{},
+			fmt.Errorf("unexpected type for bitcoinCap: expected pack.String , got %v",
```
Comment (jazg) [inline compat/v0/compat.go:78]:
```suggestion
			fmt.Errorf("unexpected type for bitcoinCap: expected pack.U64, got %v",
```

This should be `pack.U64` right?

File: compat/v0/compat.go [hunk 10]
```diff
@@ -68,6 +68,87 @@ func ShardsResponseFromState(state jsonrpc.ResponseQueryState) (ResponseQuerySha
...
+				state.State[multichain.Bitcoin].Get("gasCap").Type())
+	}
+
+	bitcoinLimit, ok := state.State[multichain.Bitcoin].Get("gasLimit").(pack.U64)
+	if !ok {
+		return ResponseQueryFees{},
+			fmt.Errorf("unexpected type for bitcoinLimit: expected pack.String , got %v",
```
Comment (jazg) [inline compat/v0/compat.go:85]:
```suggestion
			fmt.Errorf("unexpected type for bitcoinLimit: expected pack.U64, got %v",
```

File: compat/v0/compat.go [hunk 11]
```diff
@@ -68,6 +68,87 @@ func ShardsResponseFromState(state jsonrpc.ResponseQueryState) (ResponseQuerySha
...
+				state.State[multichain.Bitcoin].Get("gasLimit").Type())
+	}
+	bitcoinUnderlying := U64{Int: big.NewInt(int64(bitcoinCap.Uint64() * bitcoinLimit.Uint64()))}
+
+	zcashCap, ok := state.State[multichain.Zcash].Get("gasCap").(pack.U64)
+	if !ok {
+		return ResponseQueryFees{},
+			fmt.Errorf("unexpected type for zcashCap: expected pack.String , got %v",
```
Comment (jazg) [inline compat/v0/compat.go:93]:
```suggestion
			fmt.Errorf("unexpected type for zcashCap: expected pack.U64, got %v",
```

File: compat/v0/compat.go [hunk 12]
```diff
@@ -68,6 +68,87 @@ func ShardsResponseFromState(state jsonrpc.ResponseQueryState) (ResponseQuerySha
...
+				state.State[multichain.Zcash].Get("gasCap").Type())
+	}
+
+	zcashLimit, ok := state.State[multichain.Zcash].Get("gasLimit").(pack.U64)
+	if !ok {
+		return ResponseQueryFees{},
+			fmt.Errorf("unexpected type for zcashLimit: expected pack.String , got %v",
```
Comment (jazg) [inline compat/v0/compat.go:100]:
```suggestion
			fmt.Errorf("unexpected type for zcashLimit: expected pack.U64, got %v",
```

File: compat/v0/compat.go [hunk 13]
```diff
@@ -68,6 +68,87 @@ func ShardsResponseFromState(state jsonrpc.ResponseQueryState) (ResponseQuerySha
...
+				state.State[multichain.Zcash].Get("gasLimit").Type())
+	}
+	zcashUnderlying := U64{Int: big.NewInt(int64(zcashCap.Uint64() * zcashLimit.Uint64()))}
+
+	bitcoinCashCap, ok := state.State[multichain.BitcoinCash].Get("gasCap").(pack.U64)
+	if !ok {
+		return ResponseQueryFees{},
+			fmt.Errorf("unexpected type for bitcoinCashCap: expected pack.String , got %v",
```
Comment (jazg) [inline compat/v0/compat.go:108]:
```suggestion
			fmt.Errorf("unexpected type for bitcoinCashCap: expected pack.U64, got %v",
```

File: compat/v0/compat.go [hunk 14]
```diff
@@ -68,6 +68,87 @@ func ShardsResponseFromState(state jsonrpc.ResponseQueryState) (ResponseQuerySha
...
+				state.State[multichain.BitcoinCash].Get("gasCap").Type())
+	}
+
+	bitcoinCashLimit, ok := state.State[multichain.BitcoinCash].Get("gasLimit").(pack.U64)
+	if !ok {
+		return ResponseQueryFees{},
+			fmt.Errorf("unexpected type for bitcoinCashLimit: expected pack.String , got %v",
```
Comment (jazg) [inline compat/v0/compat.go:115]:
```suggestion
			fmt.Errorf("unexpected type for bitcoinCashLimit: expected pack.U64, got %v",
```

**Review by jazg — CHANGES_REQUESTED** [review]
Looks good overall! Just some cleanliness suggestions, but logic seems sound!

File: resolver/resolver.go (hunk 3 above)
Comment (TheAncientGoat) [reply resolver/resolver.go:309]: Fought with this for quite a while, would appreciate a better approach if you have one.

**Review by jazg — APPROVED** [review]

**Pull request merged**